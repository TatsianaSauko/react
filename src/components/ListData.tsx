import { DataAnime } from '@/types/types';
import Image from 'next/image';
interface IDataApi {
  dataApi: DataAnime[];
  handleItemClick: (item: DataAnime) => void;
}

function ListData({ dataApi, handleItemClick }: IDataApi) {
  return (
    <ul className="cards">
      {dataApi && dataApi.length ? (
        dataApi.map((item) => (
          <div key={item.mal_id} onClick={() => handleItemClick(item)}>
            <li key={item.mal_id} className="card">
              <div className="card__title">{item.title} </div>
              <Image
                src={item.images.jpg.image_url}
                alt={item.title}
                width={270}
                height={380}
              />
            </li>
          </div>
        ))
      ) : (
        <li className="no-found">Nothing found</li>
      )}
    </ul>
  );
}

export default ListData;
