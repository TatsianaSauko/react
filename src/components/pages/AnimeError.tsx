import { Link, useRouteError } from 'react-router-dom';
export default function AnimeError() {
  const error = useRouteError();
  return (
    <div className="anime-error">
      <h2>Error</h2>
      <p>{error.message}</p>
      <Link to="/">Back to the homepage</Link>
    </div>
  );
}
