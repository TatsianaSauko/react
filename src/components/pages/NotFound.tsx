import React from 'react';
import { Link } from 'react-router-dom';

export default function NotFound() {
  return (
    <>
      <div>Page not found!</div>
      <div>
        Go to the <Link to="/">Homepage</Link>.
      </div>
    </>
  );
}
