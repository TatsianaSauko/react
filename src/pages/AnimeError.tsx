import { Link, useRouteError } from 'react-router-dom';
import { IError } from '../types/types';

export default function AnimeError() {
  const error = useRouteError() as IError;
  return (
    <div className="anime-error">
      <h2>Error</h2>
      <p>{error.message}</p>
      <Link to="/">Back to the homepage</Link>
    </div>
  );
}
