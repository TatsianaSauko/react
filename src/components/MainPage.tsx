import React from 'react';
import { useState, useEffect } from 'react';
import InputField from './InputField';
import ListData from './ListData';
import ButtonError from './ButtonError';
import Loader from './Loader';
import { ItemApi } from './types/types';
import animeService from '../API/animeService';
import SelectLimit from './SelectLimit';
import Pagination from './Pagination';
import { useNavigate, Outlet } from 'react-router-dom';

const MainPage: React.FC = () => {
  const [dataInput, setDataInput] = useState<string>(
    localStorage.getItem('state') || ''
  );
  const [page, setPage] = useState<number>(1);
  const [limit, setLimit] = useState<number>(5);
  const [lastVisiblePage, setLastVisiblePage] = useState<number>(4001);
  const [dataApi, setDataApi] = useState<ItemApi[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const navigatePage = useNavigate();
  // const navigateBack = useNavigate();
  // const goBack = () => navigateBack(-1);

  const getData = async () => {
    setIsLoading(true);
    const [data, pagination] = await animeService(dataInput, page, limit);
    setDataApi(data);
    setLastVisiblePage(pagination.last_visible_page);
    setIsLoading(false);
  };

  useEffect(() => {
    getData();
    navigatePage(`?page=${page}`);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [limit, page]);

  const handleAdd = (e: React.FormEvent) => {
    e.preventDefault();
    setDataInput(dataInput.trim());
    localStorage.setItem('state', dataInput);
    getData();
  };

  const changeLimit = (data: number) => {
    setLimit(data);
    setPage(1);
  };

  return (
    <div className="app">
      <div className="main">
        <ButtonError />
        <span className="head">Anime</span>

        <InputField
          dataInput={dataInput}
          setDataInput={setDataInput}
          handleAdd={handleAdd}
        />
        <SelectLimit value={limit} changeLimit={(data) => changeLimit(data)} />
        {isLoading ? <Loader /> : <ListData prop={dataApi} />}

        <Pagination
          page={page}
          lastVisiblePage={lastVisiblePage}
          setPage={setPage}
        />
      </div>
      <Outlet />
    </div>
  );
};

export default MainPage;
