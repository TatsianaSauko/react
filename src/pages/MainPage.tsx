import React from 'react';
import { useState, useEffect } from 'react';
import InputField from '../components/InputField';
import ListData from '../components/ListData';
import ButtonError from '../components/ButtonError';
import Loader from '../components/Loader';
import { ItemApi } from '../types/types';
import animeService from '../API/animeService';
import SelectLimit from '../components/SelectLimit';
import Pagination from '../components/Pagination';
import { Outlet, useSearchParams, useNavigate } from 'react-router-dom';

const MainPage: React.FC = () => {
  const [searchParams, seatSearchParams] = useSearchParams();
  const [dataInput, setDataInput] = useState<string>(
    localStorage.getItem('state') || searchParams.get('search') || ''
  );
  const [page, setPage] = useState<number>(
    Number(searchParams.get('page')) || 1
  );
  const [limit, setLimit] = useState<number>(
    Number(searchParams.get('limit')) || 5
  );
  const [lastVisiblePage, setLastVisiblePage] = useState<number>(4001);
  const [dataApi, setDataApi] = useState<ItemApi[]>([]);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [goBack, setGoBack] = useState<boolean>(false);
  const navigate = useNavigate();
  let rootClasses = 'main';

  const getData = async () => {
    setIsLoading(true);
    const [data, pagination] = await animeService(dataInput, page, limit);
    setDataApi(data);
    setLastVisiblePage(pagination.last_visible_page);
    setIsLoading(false);
  };

  useEffect(() => {
    getData();
    seatSearchParams({ search: dataInput, limit: `${limit}`, page: `${page}` });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [limit, page]);

  const handleAdd = (e: React.FormEvent) => {
    e.preventDefault();
    setDataInput(dataInput.trim());
    localStorage.setItem('state', dataInput);
    seatSearchParams({ search: dataInput, limit: `${limit}`, page: `${page}` });
    getData();
  };

  const changeLimit = (data: number) => {
    setLimit(data);
    setPage(1);
  };
  const closePage = () => {
    if (goBack) {
      navigate(-1);
      setGoBack(false);
    }
  };

  if (goBack) {
    rootClasses = 'main overlay';
  } else {
    rootClasses = 'main';
  }

  return (
    <div className="app">
      <div onClick={closePage}>
        <div className={`${rootClasses}`}>
          <ButtonError />
          <span className="head">Anime</span>

          <InputField
            dataInput={dataInput}
            setDataInput={setDataInput}
            handleAdd={handleAdd}
          />

          <SelectLimit
            value={limit}
            changeLimit={(data) => changeLimit(data)}
          />
          {isLoading ? <Loader /> : <ListData prop={dataApi} />}

          <Pagination
            page={page}
            lastVisiblePage={lastVisiblePage}
            setPage={setPage}
          />
        </div>
      </div>

      <Outlet context={setGoBack} />
    </div>
  );
};

export default MainPage;
