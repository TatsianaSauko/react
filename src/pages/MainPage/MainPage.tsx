import React, { useState, useEffect } from 'react';
import InputField from '../../components/InputField/InputField';
import ListData from '../../components/ListData/ListData';
import ButtonError from '../../components/ButtonError/ButtonError';
import Loader from '../../components/Loader';
import SelectLimit from '../../components/SelectLimit/SelectLimit';
import Pagination from '../../components/Pagination/Pagination';
import { Outlet, useSearchParams, useNavigate } from 'react-router-dom';
import { useGetAnimeQuery } from '../../store/anime/anime.api';
import { useActions } from '../../hooks/actions';
import { useAppSelector } from '../../hooks/redux';
import AnimeError from '../AnimeError/AnimeError';

const MainPage: React.FC = () => {
  const { dataInput, limit, page } = useAppSelector((state) => state.anime);

  const { changeLastVisiblePage, changeDataApi } = useActions();

  const [, seatSearchParams] = useSearchParams();

  const [isClose, setIsClose] = useState<boolean>(false);
  const navigate = useNavigate();

  const { data, isLoading, isError } = useGetAnimeQuery({
    dataInput,
    limit,
    page,
  });

  useEffect(() => {
    if (data?.data) {
      changeDataApi(data?.data);
    }
    if (data?.lastVisiblePage) {
      changeLastVisiblePage(data?.lastVisiblePage);
    }
    seatSearchParams({ page: `${page}` });

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [data]);

  const closePage = () => {
    if (isClose) {
      navigate(-1);
      setIsClose(false);
    }
  };

  return (
    <>
      {isError ? (
        <AnimeError />
      ) : (
        <div className="app">
          <div onClick={closePage}>
            <div className={`${isClose ? 'main overlay' : 'main'}`}>
              <ButtonError />
              <span className="head">Anime</span>
              <InputField />
              <SelectLimit />
              {isLoading ? <Loader /> : <ListData />}
              <Pagination />
            </div>
          </div>
          <Outlet context={setIsClose} />
        </div>
      )}
    </>
  );
};

export default MainPage;
