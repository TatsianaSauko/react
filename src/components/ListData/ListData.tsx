import { useContext } from 'react';
import { Link } from 'react-router-dom';
import { SearchContext } from '../../pages/MainPage';

function ListData() {
  const { dataApi } = useContext(SearchContext);
  return (
    <ul className="cards">
      {dataApi.length ? (
        dataApi.map((item) => (
          <Link key={item.mal_id} to={`${item.mal_id.toString()}`}>
            <li key={item.mal_id} className="card">
              <div className="card__title">{item.title} </div>
              <img src={item.images.jpg.image_url} alt={item.title} />
            </li>
          </Link>
        ))
      ) : (
        <li className="no-found">Nothing found</li>
      )}
    </ul>
  );
}

export default ListData;
