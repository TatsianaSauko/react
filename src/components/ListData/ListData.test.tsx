import { expect, it, describe } from 'vitest';
import { render, screen } from '@testing-library/react';
import ListData from './ListData';
import { SearchContext } from '../../pages/MainPage';
import { BrowserRouter } from 'react-router-dom';

// describe('ListData component', () => {
// 	test('ListData render', () => {
// 		render(<ListData />);
// 		expect(screen.getByRole('list')).toBeDefined;

// 	  });

// 	// test('renders correct number of cards', async () => {
// 	// 	const dataApi = [
// 	// 	  { mal_id: 1, title: 'Title 1', images: { jpg: { image_url: 'url1' } } },
// 	// 	  { mal_id: 2, title: 'Title 2', images: { jpg: { image_url: 'url2' } } },
// 	// 	  { mal_id: 3, title: 'Title 3', images: { jpg: { image_url: 'url3' } } },
// 	// 	];
// 	// 		// const items = await screen.findAllByAltText('name')
// 	// 		// expect(items).toHaveLength(3)
// 	// 		const { container } = render(<ListData />);
// 	// 		const cards = container.querySelectorAll('.card');
// 	// 		console.log(cards)
// 	// 		expect(cards.length).toBe(dataApi.length);
// 	//   });
// 	test('renders correct number of cards', () => {
// 		const dataApi = [
// 		  { mal_id: 1, title: 'Title 1', images: { jpg: { image_url: 'url1' } } },
// 		  { mal_id: 2, title: 'Title 2', images: { jpg: { image_url: 'url2' } } },
// 		  { mal_id: 3, title: 'Title 3', images: { jpg: { image_url: 'url3' } } },
// 		];
// 		render(<ListData dataApi={dataApi} />);
// 		const cards = screen.getAllByRole('listitem');
// 		expect(cards).toHaveLength(dataApi.length);
// 	  });

//   test('renders "Nothing found" message if no cards are present', () => {
// 	const { getByText } = render(<ListData />);
// 	const message = getByText(/nothing found/i);
// 	expect(message).toBeInTheDocument();
//   });
// })

describe('ListData component', () => {
  it('ListData render', () => {
    render(<ListData />);
    expect(screen.getByRole('list')).toBeDefined;
  });
  it('renders correct number of cards', async () => {
    const mockData = [
      { mal_id: 1, title: 'Title 1', images: { jpg: { image_url: 'url1' } } },
      { mal_id: 2, title: 'Title 2', images: { jpg: { image_url: 'url2' } } },
    ];
    const mockSetDataInput = vi.fn();
    render(
      <BrowserRouter>
        <SearchContext.Provider
          value={{
            dataInput: '',
            setDataInput: mockSetDataInput,
            dataApi: mockData,
          }}
        >
          <ListData />
        </SearchContext.Provider>
      </BrowserRouter>
    );

    const cards = await screen.getAllByRole('listitem');
    expect(cards).toHaveLength(mockData.length);
  });

  it('renders "Nothing found" message if no cards are present', () => {
    render(
      <SearchContext.Provider
        value={{ dataInput: '', setDataInput: vi.fn(), dataApi: [] }}
      >
        <ListData />
      </SearchContext.Provider>
    );

    const message = screen.getByText(/Nothing found/i);
    expect(message).toBeInTheDocument();
  });
});
