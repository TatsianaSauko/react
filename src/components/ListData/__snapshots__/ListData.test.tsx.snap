// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListData component > renders correct number of cards 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="cards"
      >
        <a
          href="/1"
        >
          <li
            class="card"
          >
            <div
              class="card__title"
            >
              Title 1
               
            </div>
            <img
              alt="Title 1"
              src="url1"
            />
          </li>
        </a>
        <a
          href="/2"
        >
          <li
            class="card"
          >
            <div
              class="card__title"
            >
              Title 2
               
            </div>
            <img
              alt="Title 2"
              src="url2"
            />
          </li>
        </a>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="cards"
    >
      <a
        href="/1"
      >
        <li
          class="card"
        >
          <div
            class="card__title"
          >
            Title 1
             
          </div>
          <img
            alt="Title 1"
            src="url1"
          />
        </li>
      </a>
      <a
        href="/2"
      >
        <li
          class="card"
        >
          <div
            class="card__title"
          >
            Title 2
             
          </div>
          <img
            alt="Title 2"
            src="url2"
          />
        </li>
      </a>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
